File: Example_LEGO.cpp
//
// Created by pbo on 18.06.19.
//

#include <iostream>
#include <Placeholder.hpp>
#include <Operation.hpp>
#include <SummationOp.hpp>
#include <MultiplicationOp.hpp>
#include <DataInitialization.hpp>
#include <Parameter.hpp>
#include <ConvolutionOp.hpp>
#include <ReLuOp.hpp>
#include <MaxPoolOp.hpp>
#include <SoftmaxOp.hpp>
#include <CrossEntropyOp.hpp>
#include "mnist/mnist_reader.hpp"
#include <mnist/mnist_utils.hpp>

#include <lodepng.hpp>
#include <iomanip>


#include <IO.hpp>
#include <InputLayer.hpp>
#include <ConvolutionLayer.hpp>
#include <AbstractLayer.hpp>
#include <MaxPoolLayer.hpp>
#include <DenseLayer.hpp>
#include <LossLayer.hpp>
#include <LogitsLayer.hpp>
#include <NeuralNetwork.hpp>
#include <filesystem>
#include <algorithm>
#include <random>
#include <LegoDataLoader.hpp>


int main() {
//
	/*
	 * batch_size: if this is changed '#define BATCH_SIZE' in Node.hpp has to be changed as well
	 * epochs: sets the amount of epochs for training
	 * amount_batches: 'batch_size*amount_batches' gives the total amount of samples
	 * writeWeights: if set the trained Weights  are written to Source_Directory/WeightDeposit
	 * readWeights: if set (and Weights have already been Written once) weights are initialized with weights from Source_Directory/WeightDeposit
	 */
	int batch_size = 16;
	int epochs = 5;
	double learningRate = 0.0001;
	int amount_batches = 10;
	bool writeWeights = true;
	bool readWeights = false;

	std::vector<std::pair<std::string, int>> shuffledDataStrings = LegoDataLoader::shuffleData(DATA_DIR);
	DataSet legoData;
	LegoDataLoader::getData(batch_size * amount_batches, shuffledDataStrings, legoData);



	/*
 	* Create Neural Network
 	*/
	HyperParameters config(epochs, batch_size, learningRate);

	std::shared_ptr<Graph> graph = std::make_shared<Graph>();
	//Create InputLayer
	auto inputLayer = std::make_shared<InputLayer>(graph, batch_size, 160000, 4);

	//Convolutional Layer 1

	auto convolution1 = std::make_shared<ConvolutionLayer>(inputLayer, graph, ActivationType::ReLu,
			32, 8, 2, InitializationType::Xavier);

	auto maxPool1 = std::make_shared<MaxPoolLayer>(convolution1, graph, 2, 2);


	//convolutional Layer 2
	auto convolution2 = std::make_shared<ConvolutionLayer>(maxPool1, graph, ActivationType::ReLu, 64, 5, 2, InitializationType::Xavier);
	//Maxpooling
	auto maxPool2 = std::make_shared<MaxPoolLayer>(convolution2, graph, 2, 2);

	//Dense Layer 1
	auto dense1 = std::make_shared<DenseLayer>(maxPool2, graph, ActivationType::ReLu, 1024, InitializationType::Xavier);

	//Dense Layer 2
	auto dense2 = std::make_shared<DenseLayer>(dense1, graph, ActivationType::None, 16, InitializationType::Xavier);


	//Logits Layer
	auto logits = std::make_shared<LogitsLayer>(dense2, graph, 16);

	//    Cost Layer
	auto loss = std::make_shared<LossLayer>(logits, graph, LossType::CrossEntropy);

	//Create Deep Learning session
	NeuralNetwork network(graph, inputLayer, loss);

	/*
	 * Initialize Network with precalculated Weights
	 */
	if (readWeights) { network.readParameters(STORAGE, "lego_layer"); }

	/*
	 * Train the Network
	 */
	TrainingEvaluation eval = network.trainAndValidate(legoData, config);
	//		network.train(legoData,config);

	/*
	 * Write calculated Weights to Network
	 */
	if (writeWeights) { network.writeParameters(STORAGE, "lego_layer"); }


}

File: pythonApps/example_lego.py
import numpy as np
import sys
# projectDir='/mnt/c/Users/phili/Desktop/Projekte/libdl/'
projectDir='/home/pbo/CLionProjects/libdl/'
dataStorage = projectDir+"data/Storage/"

sys.path.append(projectDir+'cmake-build-release/library/bindings')


import libdl as dl

import matplotlib.pyplot as plt

pretrainedNetwork = "batch16_Samples_10"
newNetworkName = "test"
batchSize =16
epochs =5

learningRate=0.0001
amountBatches=5
dataStringPairs = dl.LegoDataLoader.shuffleData(projectDir+"data/")

data = dl.DataSet()
dl.LegoDataLoader.getData(batchSize*amountBatches,dataStringPairs,data)


graph = dl.Graph()

inputLayer = dl.InputLayer(graph,batchSize,160000,4)

convolution1 = dl.ConvolutionLayer(inputLayer,graph, dl.ActivationType.ReLu,32,8,2,dl.InitializationType.Xavier)
maxPool1 = dl.MaxPoolLayer(convolution1,graph,2,2)


convolution3 = dl.ConvolutionLayer(maxPool1,graph, dl.ActivationType.ReLu,64,5,2,dl.InitializationType.Xavier)

maxPool2 = dl.MaxPoolLayer(convolution3,graph,2,2)


dense1 = dl.DenseLayer(maxPool2,graph,dl.ActivationType.ReLu,1024,dl.InitializationType.Xavier)

dense2 = dl.DenseLayer(dense1,graph,dl.ActivationType.ReLu,16,dl.InitializationType.Xavier)

logits = dl.LogitsLayer(dense2,graph,16)

loss = dl.LossLayer(logits,graph, dl.LossType.CrossEntropy)


learningParameters = dl.HyperParameters(epochs,batchSize,learningRate)
network = dl.NeuralNetwork(graph,inputLayer,loss)
network.readParameters(dataStorage,pretrainedNetwork)
trainingEvaluation = network.trainAndValidate(data,learningParameters,1)

network.writeParameters(dataStorage,newNetworkName)


epochs = np.arange(trainingEvaluation._hyperParameters._epochs)


plt.plot()
plt.title('Training vs Validation Loss')
plt.plot(epochs, trainingEvaluation._trainingLoss, '-o', label='train')
plt.plot(epochs, trainingEvaluation._validationLoss, '-o', label='val')
plt.plot(len(epochs), 'k--')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.legend(loc='lower right')
plt.gcf().set_size_inches(15, 12)
plt.savefig(dataStorage+"loss_Graph_"+newNetworkName+".png")
plt.show()

plt.plot()
plt.title('Training vs Validation Accuracy')
plt.plot(epochs, trainingEvaluation._trainingAccuracy, '-o', label='train')
plt.plot(epochs, trainingEvaluation._validationAccuracy, '-o', label='val')
plt.plot(len(epochs), 'k--')
plt.xlabel('Epoch')
plt.ylabel('Accuracy')
plt.ylim(0,1)
plt.legend(loc='lower right')
plt.gcf().set_size_inches(15, 12)
plt.savefig(dataStorage+"accuracy_Graph_"+newNetworkName+".png")
plt.show()






